Index: web/common-web/pom.xml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- web/common-web/pom.xml	(date 1510507514000)
+++ web/common-web/pom.xml	(revision )
@@ -22,11 +22,6 @@
             <version>${project.version}</version>
         </dependency>
         <dependency>
-            <groupId>commons-fileupload</groupId>
-            <artifactId>commons-fileupload</artifactId>
-            <version>1.3.3</version>
-        </dependency>
-        <dependency>
             <groupId>javax.servlet</groupId>
             <artifactId>javax.servlet-api</artifactId>
             <version>3.1.0</version>
Index: web/upload/src/main/java/ru/javaops/masterjava/upload/UploadServlet.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- web/upload/src/main/java/ru/javaops/masterjava/upload/UploadServlet.java	(date 1510507514000)
+++ web/upload/src/main/java/ru/javaops/masterjava/upload/UploadServlet.java	(revision )
@@ -1,23 +1,23 @@
 package ru.javaops.masterjava.upload;
 
-import org.apache.commons.fileupload.FileItemIterator;
-import org.apache.commons.fileupload.FileItemStream;
-import org.apache.commons.fileupload.servlet.ServletFileUpload;
 import org.thymeleaf.context.WebContext;
 import ru.javaops.masterjava.model.User;
 
 import javax.servlet.ServletException;
+import javax.servlet.annotation.MultipartConfig;
 import javax.servlet.annotation.WebServlet;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
+import javax.servlet.http.Part;
 import java.io.IOException;
 import java.io.InputStream;
 import java.util.List;
 
 import static ru.javaops.masterjava.common.web.ThymeleafListener.engine;
 
-@WebServlet("/")
+@WebServlet(urlPatterns = "/", loadOnStartup = 1)
+@MultipartConfig(fileSizeThreshold = 1024 * 1024 * 10) //10 MB in memory limit
 public class UploadServlet extends HttpServlet {
 
     private final UserProcessor userProcessor = new UserProcessor();
@@ -30,23 +30,18 @@
 
     @Override
     protected void doPost(HttpServletRequest req, HttpServletResponse resp) throws ServletException, IOException {
-        final ServletFileUpload upload = new ServletFileUpload();
         final WebContext webContext = new WebContext(req, resp, req.getServletContext(), req.getLocale());
 
         try {
-//            https://commons.apache.org/proper/commons-fileupload/streaming.html
-
-            final FileItemIterator itemIterator = upload.getItemIterator(req);
-            while (itemIterator.hasNext()) { //expect that it's only one file
-                FileItemStream fileItemStream = itemIterator.next();
-                if (!fileItemStream.isFormField()) {
-                    try (InputStream is = fileItemStream.openStream()) {
-                        List<User> users = userProcessor.process(is);
-                        webContext.setVariable("users", users);
-                        engine.process("result", webContext, resp.getWriter());
-                    }
-                    break;
-                }
+//            http://docs.oracle.com/javaee/6/tutorial/doc/glraq.html
+            Part filePart = req.getPart("fileToUpload");
+            if (filePart.getSize() == 0) {
+                throw new IllegalStateException("Upload file have not been selected");
+            }
+            try (InputStream is = filePart.getInputStream()) {
+                List<User> users = userProcessor.process(is);
+                webContext.setVariable("users", users);
+                engine.process("result", webContext, resp.getWriter());
             }
         } catch (Exception e) {
             webContext.setVariable("exception", e);
